ath79: consolidate Netgear EX7300 series images

looi opened the pull request:
This change consolidates Netgear EX7300 series devices into two images
corresponding to devices that share the same [manufacturer firmware](https://kb.netgear.com/000064217/EX6250-EX6400v2-EX6410-EX6420-EX7300v2-EX7320-Firmware-Version-1-0-0-146)
image. Similar to the manufacturer firmware, the actual device model is
detected at runtime. The logic is taken from the netgear GPL dumps in a
file called generate_board_conf.sh.

Hardware details for EX7300 v2 variants
---------------------------------------
SoC: QCN5502
Flash: 16 MiB
RAM: 128 MiB
Ethernet: 1 gigabit port
Wireless 2.4GHz (currently unsupported due to lack of ath9k support):
- EX6250 / EX6400 v2 / EX6410 / EX6420: QCN5502 3x3
- EX7300 v2 / EX7320: QCN5502 4x4

Wireless 5GHz:
- EX6250: QCA9986 3x3 (detected by ath10k as QCA9984 3x3)
- EX6400 v2 / EX6410 / EX6420 / EX7300 v2 / EX7320: QCA9984 4x4

alexconstsh commented:
Hi. I am curious, why do you use different names for the SoC in different places? You have `qcn5502` everywhere except in DTS `compatible` field where it's `qcn5500`.

looi commented:
> Hi. I am curious, why do you use different names for the SoC in different places? You have `qcn5502` everywhere except in DTS `compatible` field where it's `qcn5500`.

This was trying to indicate qcn550x, following the pattern where e.g. `qca,qca9560` indicates QCA956x, but not sure what is the convention here

alexconstsh commented:
> This was trying to indicate qcn550x, following the pattern where e.g. `qca,qca9560` indicates QCA956x, but not sure what is the convention here

Oh... I thought this field referred to a specific SoC because I remember the original device tree documentation (separate from Linux and OpenWrt) advising against wildcards.

hauke commented:
I agree with @alexconstsh  and you prefer something like this:
``
compatible = "netgear,ex6250-v1", "qca,qcn5502", "qca,qcn5500";
``
Why did you add `qca,qca9560`?

looi commented:
> I agree with @alexconstsh and you prefer something like this: `compatible = "netgear,ex6250-v1", "qca,qcn5502", "qca,qcn5500", "qca,qca9560";`

Sure, that sounds good. I split this into 2 commits, first one that adds `qca,qcn5502` to all qcn5502 devices, and second one that adds support for the new devices.

DragonBluep commented on target/linux/ath79/dts/qcn5502_netgear_ex7300-v2.dtsi:
```diff
@@ -0,0 +1,252 @@
+};
+
+&pll {
+	clocks = <&extosc>;
```
Please drop this node, it has already been defined in qca956x.dtsi.

looi replied:
Done

looi commented:
I've also updated this PR to be based on the latest master

hauke commented on target/linux/ath79/dts/qcn5502_netgear_ex7300-v2.dts:
```diff
@@ -1,255 +1,8 @@
-
-&pll {
-	clocks = <&extosc>;
-};
```
Why did you remove the `pll`node?

DragonBluep replied:
This is redundant. We have already defined it in `qca956x.dtsi`. I did a lot of cleaning in #10925.
https://github.com/openwrt/openwrt/blob/cb24be47ffda12616a0670e9d29a96f9530a19b8/target/linux/ath79/dts/qca956x.dtsi#L97

hauke commented:
The new packages have the same device tree except for the compatible string and the name. They also have the same packages. Do we really need separate images for them?

DragonBluep commented:
https://github.com/openwrt/openwrt/blob/cb24be47ffda12616a0670e9d29a96f9530a19b8/include/image.mk#L339-L348
Now we only support 3 alternative devices. Maybe we need to add extra `ALT3` and `ALT4`?

looi commented:
> The new packages have the same device tree except for the compatible string and the name. They also have the same packages. Do we really need separate images for them?

Not sure. Based on the discussion [here](https://github.com/openwrt/openwrt/pull/1379#issuecomment-480438673) for the EX6400 v1 / EX7300 v1, it seemed like it's more desirable to have separate images. But maybe it would make sense to have the same image for all of these identical devices? This would require support for more alternative devices (like @DragonBluep mentioned). Ideally, these alternate names could also be shown as the "Model" in the web UI.

looi committed:
build: add ALT3 and ALT4 vendor/model/variant

This is needed for the Netgear EX7300 series v2.

Signed-off-by: Wenli Looi <wlooi@ucalgary.ca>

looi committed:
ath79: consolidate Netgear EX7300 series images

This change consolidates Netgear EX7300 series devices into two images
corresponding to devices that share the same manufacturer firmware
image. Similar to the manufacturer firmware, the actual device model is
detected at runtime. The logic is taken from the netgear GPL dumps in a
file called generate_board_conf.sh.

Hardware details for EX7300 v2 variants
---------------------------------------
SoC: QCN5502
Flash: 16 MiB
RAM: 128 MiB
Ethernet: 1 gigabit port
Wireless 2.4GHz (currently unsupported due to lack of ath9k support):
- EX6250 / EX6400 v2 / EX6410 / EX6420: QCN5502 3x3
- EX7300 v2 / EX7320: QCN5502 4x4
Wireless 5GHz:
- EX6250: QCA9986 3x3 (detected by ath10k as QCA9984 3x3)
- EX6400 v2 / EX6410 / EX6420 / EX7300 v2 / EX7320: QCA9984 4x4

Signed-off-by: Wenli Looi <wlooi@ucalgary.ca>

looi commented:
I've changed this PR to instead consolidate all of the EX7300 series devices into two images, and perform detection of the device name at runtime. I added extra alternate device names to accommodate all of the devices that share a single image.

Let me know what you think @hauke @DragonBluep

looi commented:
Updated to latest master. This change will allow all EX7300 series devices to share two images while displaying the correct model number in LuCI and in the [firmware selector](https://firmware-selector.openwrt.org), but let me know if something else is preferred

hauke commented on target/linux/ath79/image/generic.mk:
```diff
@@ -1745,24 +1748,24 @@ define Device/netgear_ex6400_ex7300
   IMAGE/factory.img := $$(IMAGE/default) | netgear-dni | check-size
   DEVICE_PACKAGES := kmod-ath10k-ct ath10k-firmware-qca99x0-ct
```
Please add:
````
SUPPORTED_DEVICES += netgear_ex6400
````
This will add a tag to the image which tells swupgrade that the new `netgear_ex7300` image can be installed over an image with the name `netgear_ex6400` without problems. 

Without this change swupgrade will reject the new firmware when you want to upgrade it over an existing image.

Please try if this really work too.

looi replied:
Done. I had to use a slightly different line `SUPPORTED_DEVICES += netgear,ex6400` and I confirmed that sysupgrade/LuCI does not complain when upgrading.

hauke commented on target/linux/ath79/image/generic.mk:
```diff
@@ -1732,9 +1732,12 @@ define Device/nec_wg800hp
+  DEVICE_MODEL := EX7300
+  DEVICE_ALT0_VENDOR := NETGEAR
+  DEVICE_ALT0_MODEL := EX6400
   UIMAGE_MAGIC := 0x27051956
```
`0x27051956` is the default u-boot magic. Please remove this line.

looi replied:
Done

hauke commented on target/linux/ath79/image/generic.mk:
```diff
@@ -1745,24 +1748,24 @@ define Device/netgear_ex6400_ex7300
+  DEVICE_ALT3_MODEL := EX6420
+  DEVICE_ALT4_VENDOR := NETGEAR
+  DEVICE_ALT4_MODEL := EX7320
   UIMAGE_MAGIC := 0x27051956
```
`0x27051956` is the default u-boot magic. Please remove this line.

looi replied:
Done

DragonBluep commented on target/linux/ath79/image/generic.mk:
```diff
@@ -1732,9 +1732,12 @@ define Device/nec_wg800hp
 
-define Device/netgear_ex6400_ex7300
+define Device/netgear_ex7300
   $(Device/netgear_generic)
```
` $(Device/netgear_generic)` recipe is used for some very old AR724x NETGEAR devices. Could you please try to see if it can work after removing it?

looi replied:
Done. Yes, it works, after adding `DEVICE_VENDOR` and `IMAGES`. I have confirmed that upgrading both from sysupgrade and the factory web UI works for both EX7300 and EX7300 v2 series.

DragonBluep replied:
Happy to hear it works. 

By the way, many developers are working to convert wireless calibration data into NVMEM format. If you are interested, maybe you can try to convert it. Just some information, not something that must be done now. Refer to https://github.com/openwrt/openwrt/commit/3e3e78de1103cb0c5af4cc229152236758c37901

DragonBluep replied:
No, only EX7300 v2 now. I see that the EX7300 still uses `mtd-cal-data=<&caldata 0x1000>;`.

looi replied:
Ah, I see. I've added a commit to convert it now. Checked that the wifi mac addresses are as expected.

looi committed:
ath79: convert Netgear EX7300 caldata to nvmem

Transition to specify caldata in the DTS.

Signed-off-by: Wenli Looi <wlooi@ucalgary.ca>
